### PR TITLE
[CNVS Upgrade: Stylelint 1 of 3] Add styles linter

### PR DESCRIPTION
**Pull request by jfurrow**
This PR introduces [stylelint](https://github.com/stylelint/stylelint) and a bunch of rules! A definition and example of every rule can be found [here](https://github.com/stylelint/stylelint/blob/master/docs/user-guide/rules.md).

Let me know if you have any questions or concerns. I'd like to get the team's approval on these rules before fixing all of the warnings. I quickly tested [stylefmt](https://github.com/morishitter/stylefmt) to automatically format our LESS files to adhere to these rules and it seems to work well!

The webpack configuration modifications will display any lint warning each time the LESS is changed.

### CODE REVIEW

File: npm-shrinkwrap.json [hunk 1]
```diff
@@ -1605,7 +1683,7 @@
         "isarray": {
           "version": "0.0.1",
           "from": "isarray@0.0.1",
-          "resolved": "https://registry.npmjs.org/isarray/-/isarray-0.0.1.tgz"
+          "resolved": "http://registry.npmjs.org/isarray/-/isarray-0.0.1.tgz"
```
Comment (mlunoe) [inline npm-shrinkwrap.json]: These should not be changed to `http`

File: npm-shrinkwrap.json [hunk 2]
```diff
@@ -1647,7 +1725,7 @@
...
```
Comment (mlunoe) [inline npm-shrinkwrap.json]: Same here

File: npm-shrinkwrap.json [hunk 3]
```diff
@@ -1679,7 +1757,7 @@
...
```
Comment (mlunoe) [inline npm-shrinkwrap.json]: Same here

File: npm-shrinkwrap.json [hunk 4]
```diff
@@ -1711,7 +1789,7 @@
...
```
Comment (mlunoe) [inline npm-shrinkwrap.json]: Same here

File: npm-shrinkwrap.json [hunk 5]
```diff
@@ -2824,7 +2939,7 @@
...
```
Comment (mlunoe) [inline npm-shrinkwrap.json]: Again here. I'll let you take care of the rest

File: npm-shrinkwrap.json (hunk 1 above)
Comment (jfurrow) [inline npm-shrinkwrap.json]: So do we manually change all of these to `https`? This was generated automatically by running `npm run shrinkwrap`

Comment (mlunoe) [inline npm-shrinkwrap.json]: Yes :(
I get the same result (for some reason, but @rcorral's setup gives him https for all. Not sure what is different with our environments)

File: .stylelintrc [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  "rules": {
+    "at-rule-empty-line-before": "always",
+    "at-rule-name-case": "lower",
+    "at-rule-name-space-after": "always",
+    "at-rule-no-unknown": true,
+    "at-rule-no-vendor-prefix": true,
+    "at-rule-semicolon-newline-after": true,
+    "block-closing-brace-empty-line-before": "never",
+    "block-closing-brace-newline-after": "always",
+    "block-closing-brace-newline-before": "always",
```
Comment (mlunoe) [inline .stylelintrc]: Do we want this? And isn't this contradicting this rule? `block-closing-brace-empty-line-before` Maybe I am misunderstanding them

File: .stylelintrc [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    "block-no-empty": true,
+    "block-no-single-line": true,
+    "block-opening-brace-newline-after": "always",
```
Comment (mlunoe) [inline .stylelintrc]: Same here, do we want a newline after every opening brace?

File: .stylelintrc [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    "block-opening-brace-space-before": "always",
+    "color-hex-length": "short",
```
Comment (mlunoe) [inline .stylelintrc]: Should we also add `"color-hex-case": "lower"`?

File: .stylelintrc [hunk 9]
```diff
@@ -0,0 +1,109 @@
+{
+  "rules": {
```
Comment (mlunoe) [inline .stylelintrc]: Is there a reason why we don't extend the standard: https://github.com/stylelint/stylelint-config-standard

```sh
npm install stylelint-config-standard --save-dev
```

```js
{
  "extends": "stylelint-config-standard",
  "rules": {
    ...
  }
}
```

File: .stylelintrc [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    "shorthand-property-no-redundant-values": true,
+    "string-no-newline": true,
+    "string-quotes": "single",
+    "stylelint-disable-reason": "always-before",
+    "unit-case": "lower",
+    "unit-no-unknown": true,
+    "unit-whitelist": ["em", "rem", "%", "px", "s", "ms"],
+    "value-keyword-case": "lower",
+    "value-list-comma-space-after": "always",
+    "value-list-max-empty-lines": 0
```
Comment (mlunoe) [inline .stylelintrc]: I think it looks good. Could we add the linting to the CI (maybe separate PR) and fix the rules for our styles? That would also make it _a lot_ easier to understand the rules.

File: .stylelintrc (hunk 6 above)
Comment (jfurrow) [inline .stylelintrc]: I definitely want these rules, as it greatly increases legibility. And no, it doesn't contradict the `empty-line` rule. Check out the rule examples:
https://github.com/stylelint/stylelint/blob/master/src/rules/block-closing-brace-newline-after/README.md
https://github.com/stylelint/stylelint/blob/master/src/rules/block-closing-brace-newline-before/README.md
https://github.com/stylelint/stylelint/blob/master/src/rules/block-closing-brace-empty-line-before/README.md

File: .stylelintrc (hunk 9 above)
Comment (jfurrow) [inline .stylelintrc]: I wasn't aware this existed, but let's extend it! I'll have to go through this again tomorrow to see what we need to override or append.

File: .stylelintrc (hunk 7 above)
Comment (jfurrow) [inline .stylelintrc]: Yes :)

File: .stylelintrc (hunk 10 above)
Comment (jfurrow) [inline .stylelintrc]: As I mentioned in the description, I intend to fix the rules for all styles on the `cnvs-upgrade/master` branch, but I want to make sure the team agrees with the ruleset first. The lint warning fixes will be a separate PR.

I'd love for linting in CI also, but that's a separate task.

File: .stylelintrc (hunk 8 above)
Comment (jfurrow) [inline .stylelintrc]: Let's do it!

File: npm-shrinkwrap.json (hunk 1 above)
Comment (jfurrow) [inline npm-shrinkwrap.json]: Alright, I'll fix tomorrow. Perhaps I'll add it to the "fix shrinkwrap" gulp task!

Comment (jfurrow) [inline npm-shrinkwrap.json]: Done!

File: npm-shrinkwrap.json (hunk 2 above)
Comment (jfurrow) [inline npm-shrinkwrap.json]: Done!

File: npm-shrinkwrap.json (hunk 3 above)
Comment (jfurrow) [inline npm-shrinkwrap.json]: Done!

File: npm-shrinkwrap.json (hunk 4 above)
Comment (jfurrow) [inline npm-shrinkwrap.json]: Done!

File: npm-shrinkwrap.json (hunk 5 above)
Comment (jfurrow) [inline npm-shrinkwrap.json]: Done!

File: .stylelintrc (hunk 9 above)
Comment (mlunoe) [inline .stylelintrc]: Sounds good!

File: .stylelintrc (hunk 6 above)
Comment (mlunoe) [inline .stylelintrc]: Arh! I didn't see the examples! Sweet! Thank you!

File: .stylelintrc (hunk 7 above)
Comment (mlunoe) [inline .stylelintrc]: I see, thanks

File: .stylelintrc (hunk 10 above)
Comment (mlunoe) [inline .stylelintrc]: Yup yup yup! Separate tasks!
Just needed some examples which was in the document you provided, I just didn't click into them, so thanks for showing me!

File: .stylelintrc (hunk 9 above)
Comment (jfurrow) [inline .stylelintrc]: Done!

File: .stylelintrc (hunk 6 above)
Comment (jfurrow) [inline .stylelintrc]: Yay!

File: .stylelintrc (hunk 8 above)
Comment (jfurrow) [inline .stylelintrc]: Done! This is actually covered in the default that we're now extending

File: .stylelintrc (hunk 10 above)
Comment (jfurrow) [inline .stylelintrc]: Cool!

**Comment by jfurrow** [comment]
@mlunoe I think all of the concerns are addressed!

File: .stylelintrc [hunk 11]
```diff
@@ -0,0 +1,51 @@
+{
+  "extends": "stylelint-config-standard",
```
Comment (mlunoe) [inline .stylelintrc]: ❤️

**Pull request merged**